### PR TITLE
bug: Check if window is available before use 

### PR DESCRIPTION
**Pull request by JeetChaudhari**
Checked if window is available before use in app bar, slider component and ripple service, However window is used in following files as well but I could not fix all. There is closed issue for this problem at [https://github.com/trimox/angular-mdc-web/issues/33](https://github.com/trimox/angular-mdc-web/issues/33)

- [x] app-bar.ts
- [x] slider.ts
- [x] ripple.service.ts
- [x] notched-outline.ts
- [x] tab-bar.ts
- [x] tab-scroller.ts

document uses without check at following files
- [x] menu.ts
- [x] drawer.ts
- [x] ripple.service.ts
- [x] snackbar.component.ts
- [x] list.ts
- [x] menu.ts
- [x] portal.service.ts
- [x] tab-scroller.ts

This is my 1st PR on Github, so if anything is wrong with this please let me know.

### CODE REVIEW

**Comment by trimox** [comment]
@JeetChaudhari Sorry about the broken server rendering for those components! Wow, original issue was from the beginning of the repo, memories.. 😄 

Much appreciate the work! 💎 and congrats on your first PR 🥇 on GitHub!

Soon as you're ready for review or hit a snag, just ping me. We'll get this merged for a v0.38.x release.

**Comment by JeetChaudhari** [comment]
@trimox I have fixed all problems which caused window is not defined errors, Please review these changes and If you need anything changed in this then do let me know. Also, I noticed on Angular SSR some other random errors also occur with all these components, I will soon add issues for them as well as set up the repo to reproduce those random errors.

File: packages/list/list.ts [hunk 1]
```diff
@@ -158,8 +158,11 @@ export class MdcList implements AfterContentInit, OnDestroy {
 
   private _mdcAdapter: MDCListAdapter = {
     getListItemCount: () => this._listItems.length,
-    getFocusedElementIndex: () =>
-      this._listItems.toArray().findIndex((_) => _.elementRef.nativeElement === document.activeElement),
+    getFocusedElementIndex: () => {
+      if (!isBrowser()) { return false; }
```
Comment (trimox) [inline packages/list/list.ts]: Return `-1` instead of boolean

File: packages/notched-outline/notched-outline.ts [hunk 2]
```diff
@@ -36,8 +38,11 @@ export class MdcNotchedOutline implements OnInit, OnDestroy {
     addClass: (className: string) => this._notchOutline.nativeElement.classList.add(className),
     removeClass: (className: string) => this._notchOutline.nativeElement.classList.remove(className),
     setOutlinePathAttr: (value: string) => this._svgpath.nativeElement.setAttribute('d', value),
-    getIdleOutlineStyleValue: (propertyName: string) =>
-      window.getComputedStyle(this._notchIdle.nativeElement).getPropertyValue(propertyName)
+    getIdleOutlineStyleValue: (propertyName: string) => {
+      if (!isBrowser()) { return false; }
```
Comment (trimox) [inline packages/notched-outline/notched-outline.ts]: Let's rewrite this as,
```ts
getIdleOutlineStyleValue: (propertyName: string) =>
  isBrowser() ? '' : window.getComputedStyle(this._notchIdle.nativeElement).getPropertyValue(propertyName),
```

File: packages/tab-scroller/tab-scroller.ts [hunk 3]
```diff
@@ -53,14 +54,22 @@ export class MdcTabScroller implements AfterViewInit, OnDestroy {
...
+      return window.getComputedStyle(this.content.nativeElement).getPropertyValue(propName);
+    },
     setScrollAreaScrollLeft: (scrollX: number) => this.area.nativeElement.scrollLeft = scrollX,
     getScrollAreaScrollLeft: () => this.area.nativeElement.scrollLeft,
     getScrollContentOffsetWidth: () => this.content.nativeElement.offsetWidth,
     getScrollAreaOffsetWidth: () => this.area.nativeElement.offsetWidth,
     computeScrollAreaClientRect: () => this.area.nativeElement.getBoundingClientRect(),
     computeScrollContentClientRect: () => this.content.nativeElement.getBoundingClientRect(),
-    computeHorizontalScrollbarHeight: () => util.computeHorizontalScrollbarHeight(document)
+    computeHorizontalScrollbarHeight: () => {
```
Comment (trimox) [inline packages/tab-scroller/tab-scroller.ts]: Minor refactor,

```ts
computeHorizontalScrollbarHeight: () =>
  isBrowser() ? 0 : util.computeHorizontalScrollbarHeight(document),
```

**Comment by JeetChaudhari** [comment]
@trimox Thanks for reviewing the changes, I have made corrections that you asked for. Thank You

**Comment by trimox** [comment]
Thanks! 💎 I'll finish reviewing today, and we'll give it a test run.

File: packages/portal/portal.service.ts [hunk 4]
```diff
@@ -19,7 +20,7 @@ export class MdcPortalService {
     private _appRef: ApplicationRef,
     private _injector: Injector) { }
 
-  createComponentRef<T>(portal: ComponentType<T>, parent: HTMLElement = document.body): ComponentRef<T> {
+  createComponentRef<T>(portal: ComponentType<T>, parent: HTMLElement | false = isBrowser() ? document.body : false): ComponentRef<T> {
```
Comment (trimox) [inline packages/portal/portal.service.ts]: Is snackbar or dialog usable at all with server rendering?

File: packages/tab-bar/tab-bar.ts [hunk 5]
```diff
@@ -108,7 +108,11 @@ export class MdcTabBar implements AfterContentInit, OnDestroy {
     getScrollPosition: () => this.tabScroller.getScrollPosition(),
     getScrollContentWidth: () => this.tabScroller.getScrollContentWidth(),
     getOffsetWidth: () => this._getHostElement().offsetWidth,
-    isRTL: () => window.getComputedStyle(this._getHostElement()).getPropertyValue('direction') === 'rtl',
+    isRTL: () => {
```
Comment (trimox) [inline packages/tab-bar/tab-bar.ts]:
```ts
isRtl: () => isBrowser() ? window.getComputedStyle(this._getHostElement()).getPropertyValue('direction') : false,
```

File: packages/notched-outline/notched-outline.ts [hunk 6]
```diff
@@ -37,7 +39,7 @@ export class MdcNotchedOutline implements OnInit, OnDestroy {
     removeClass: (className: string) => this._notchOutline.nativeElement.classList.remove(className),
     setOutlinePathAttr: (value: string) => this._svgpath.nativeElement.setAttribute('d', value),
     getIdleOutlineStyleValue: (propertyName: string) =>
-      window.getComputedStyle(this._notchIdle.nativeElement).getPropertyValue(propertyName)
+      isBrowser() ? '' : window.getComputedStyle(this._notchIdle.nativeElement).getPropertyValue(propertyName),
```
Comment (trimox) [inline packages/notched-outline/notched-outline.ts]: Return value is backwards
```ts
isBrowser() ? window.getComputedStyle(this._notchIdle.nativeElement).getPropertyValue(propertyName) : '',
```

File: packages/tab-scroller/tab-scroller.ts [hunk 7]
```diff
@@ -53,14 +54,19 @@ export class MdcTabScroller implements AfterViewInit, OnDestroy {
     addScrollAreaClass: (className: string) => this.area.nativeElement.classList.add(className),
     setScrollAreaStyleProperty: (propName: string, value: string) => this.area.nativeElement.style.setProperty(propName, value),
     setScrollContentStyleProperty: (propName: string, value: string) => this.content.nativeElement.style.setProperty(propName, value),
-    getScrollContentStyleValue: (propName: string) => window.getComputedStyle(this.content.nativeElement).getPropertyValue(propName),
+    getScrollContentStyleValue: (propName: string) => {
```
Comment (trimox) [inline packages/tab-scroller/tab-scroller.ts]: Needs to return a string
```ts
isBrowser() ? window.getComputedStyle(this.content.nativeElement).getPropertyValue(propName) : '',
```

File: packages/tab-scroller/tab-scroller.ts [hunk 8]
```diff
@@ -53,14 +54,19 @@ export class MdcTabScroller implements AfterViewInit, OnDestroy {
...
+      return window.getComputedStyle(this.content.nativeElement).getPropertyValue(propName);
+    },
     setScrollAreaScrollLeft: (scrollX: number) => this.area.nativeElement.scrollLeft = scrollX,
     getScrollAreaScrollLeft: () => this.area.nativeElement.scrollLeft,
     getScrollContentOffsetWidth: () => this.content.nativeElement.offsetWidth,
     getScrollAreaOffsetWidth: () => this.area.nativeElement.offsetWidth,
     computeScrollAreaClientRect: () => this.area.nativeElement.getBoundingClientRect(),
     computeScrollContentClientRect: () => this.content.nativeElement.getBoundingClientRect(),
-    computeHorizontalScrollbarHeight: () => util.computeHorizontalScrollbarHeight(document)
+    computeHorizontalScrollbarHeight: () =>
```
Comment (trimox) [inline packages/tab-scroller/tab-scroller.ts]: Return value is backwards for browsers
```ts
isBrowser() ? util.computeHorizontalScrollbarHeight(document) : 0,
```

**Comment by JeetChaudhari** [comment]
@trimox Thanks for pointing out changes, I will get this tomorrow, it would be soon night here in India.

**Comment by trimox** [comment]
It's no problem! Sorry for delay on review, been tied up with upcoming MDC v0.39.0 changes..

**Comment by trimox** [comment]
Also, I'd recommend holding off on merging master if you're testing still. I began merging v0.39.0 features today.

File: packages/tab-bar/tab-bar.ts (hunk 5 above)
Comment (JeetChaudhari) [reply packages/tab-bar/tab-bar.ts]: Are we changing function name from isRTL to isRtl?

File: packages/portal/portal.service.ts (hunk 4 above)
Comment (JeetChaudhari) [reply packages/portal/portal.service.ts]: Yes, I have used snack bar as well as dialog in my project which uses Angular SSR.

File: packages/notched-outline/notched-outline.ts (hunk 6 above)
Comment (JeetChaudhari) [reply packages/notched-outline/notched-outline.ts]: Fixed this mistake.

File: packages/tab-scroller/tab-scroller.ts (hunk 7 above)
Comment (JeetChaudhari) [reply packages/tab-scroller/tab-scroller.ts]: Fixed.

File: packages/tab-scroller/tab-scroller.ts (hunk 8 above)
Comment (JeetChaudhari) [reply packages/tab-scroller/tab-scroller.ts]: Fixed.

File: packages/tab-bar/tab-bar.ts (hunk 5 above)
Comment (trimox) [reply packages/tab-bar/tab-bar.ts]: Good catch! Should be `isRTL`, and returns boolean.

```ts
isRTL: () => isBrowser() ? window.getComputedStyle(this._getHostElement()).getPropertyValue('direction') === 'rtl' : false,
```

File: packages/portal/portal.service.ts (hunk 4 above)
Comment (trimox) [reply packages/portal/portal.service.ts]: Great! Glad that's working.

File: packages/tab-bar/tab-bar.ts (hunk 5 above)
Comment (JeetChaudhari) [reply packages/tab-bar/tab-bar.ts]: Fixed

**Comment by JeetChaudhari** [comment]
@trimox All required changes completed.

**Comment by trimox** [comment]
LGTM! Don't worry about the merge conflicts, I'll take a look.

Also, I've discovered the root cause of the random errors you mentioned on some components. I'll create an issue tomorrow, but it's related to the way TypeScript preserves type information of methods with decorators (e.g.: `@HostBindings` and `@HostListener`).

**Comment by JeetChaudhari** [comment]
Thank You

**Comment by trimox** [comment]
I'm slowly but surely incorporating changes reporting conflicts with the major changes in v0.39.1. We're getting there..

**Comment by trimox** [comment]
And implementing the additional recommended fixes for other areas reporting issues for server rendering - reference #1268

**Comment by JeetChaudhari** [comment]
@trimox No issues. For #1268 I do not understand what actual changes you intend to do by the configurable directive, If you can show me an example of one component, I am willing to help. Actually, I was a PHP developer since 2012, but from last year I started using only javascript lang in my company for so many issues and save costing instead of multiple programming langs, so still don't have much experience with all this javascript.

**Comment by trimox** [comment]
@JeetChaudhari Don't fret about these merge conflicts, they are expected since the upcoming v0.39.x release had a ton of adapter improvements upstream. I've tried to incorporate your PR fixes as I progressed with the updates. So far it's looking good!

Directive configuration example,
https://github.com/trimox/angular-mdc-web/blob/master/packages/menu/menu.ts#L74

**Comment by trimox** [comment]
Phew, since v0.39.x got published I'll work on merging into v0.38.2 tag as separate branch..

**Comment by JeetChaudhari** [comment]
@trimox Can we please merge this into master and later sort out other issues (#1268 )? Merging this would make SSR build work for at least other components which are not affected by #1268 

Thank You

**Comment by trimox** [comment]
Closing. Created #1381 off this PR

**Pull request closed without merging**